dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,337 @@
+            sbStore.append(rightBracket);
+        }
+        // separator
+        sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+
+        return sbStore.toString();
+    }
+
+    protected String visualizeSpecialBlockHelper(SpecialBlock sbb, String separator) {
+        String specialBlock = "";
+        switch (sbb.getSpecialType()) {
+            case ENTRY:
+                specialBlock = "<entry>" + separator;
+                break;
+            case EXIT:
+                specialBlock = "<exit>" + separator;
+                break;
+            case EXCEPTIONAL_EXIT:
+                specialBlock = "<exceptional-exit>" + separator;
+                break;
+        }
+        return specialBlock;
+    }
+
+    protected Node getLastNode(Block bb) {
+        Node lastNode;
+        switch (bb.getType()) {
+            case REGULAR_BLOCK:
+                List<Node> blockContents = ((RegularBlock) bb).getContents();
+                lastNode = blockContents.get(blockContents.size() - 1);
+                break;
+            case EXCEPTION_BLOCK:
+                lastNode = ((ExceptionBlock) bb).getNode();
+                break;
+            default:
+                lastNode = null;
+        }
+        return lastNode;
+    }
+
+    @Override
+    public String visualizeBlockNode(Node t, @Nullable Analysis<A, S, T> analysis) {
+        StringBuilder sbBlockNode = new StringBuilder();
+        sbBlockNode
+                .append(prepareString(t.toString()))
+                .append("   [ ")
+                .append(prepareNodeType(t))
+                .append(" ]");
+        if (analysis != null) {
+            A value = analysis.getValue(t);
+            if (value != null) {
+                sbBlockNode.append("    > ").append(prepareString(value.toString()));
+            }
+        }
+        return sbBlockNode.toString();
+    }
+
+    /**
+     * Generate the order of processing blocks.
+     *
+     * @param cfg the current control flow graph
+     */
+    IdentityHashMap<Block, List<Integer>> getProcessOrder(ControlFlowGraph cfg) {
+        IdentityHashMap<Block, List<Integer>> depthFirstOrder = new IdentityHashMap<>();
+        int count = 1;
+        for (Block b : cfg.getDepthFirstOrderedBlocks()) {
+            depthFirstOrder.computeIfAbsent(b, k -> new ArrayList<>());
+            depthFirstOrder.get(b).add(count++);
+        }
+        return depthFirstOrder;
+    }
+
+    @Override
+    public String visualizeStore(S store) {
+        return store.visualize(this);
+    }
+
+    @Override
+    public String visualizeStoreFooter() {
+        return ")";
+    }
+
+    protected String prepareNodeType(Node t) {
+        String name = t.getClass().getSimpleName();
+        return name.replace("Node", "");
+    }
+
+    /**
+     * Escape double quotes.
+     *
+     * @param s the String to be processed.
+     */
+    protected String prepareString(String s) {
```

Comment:
This is only necessary for strings that are used within a dot string, e.g. something that will be within a dot label.
Maybe this should be another abstract method and the DOT version overrides it.
Or this version does nothing and the dot version overrides that behavior.

Reply:
I was wondering if we can remove this method and simply use `toStringEscapeDoubleQuotes`.

```java
    protected String prepareString(String s) {
        return s.replace("\"", "\\\"");
    }

    protected String escapeDoubleQuotes(final String str) {
        return str.replace("\"", "\\\"");
    }

    protected String toStringEscapeDoubleQuotes(final Object obj) {
        return escapeDoubleQuotes(String.valueOf(obj));
    }
```
It seems good for us, the bodies of `prepareString` and `escapeDoubleQuotes` are the same, and we can just write `toStringEscapeDoubleQuotes(t)` instead of  `prepareString(t.toString())`. But, I am not sure if `final Object obj` will be an issue, since `prepareString` are not forced to use `final`.

I had a simple test and it seems worked properly. But, I still doubt.

Reply:
Are you proposing new methods or renaming existing methods? I'm confused what exactly you're suggesting.
The `final Object` prevents re-assignment within that method body. It has no impact on callers of the method.

Reply:
In case you're suggesting new methods, I would overload the name. Then overload resolution will either pick the `String` version or the `Object` version and will do the right thing for each.

```java
    protected String escapeDoubleQuotes(final String str) {
        return str.replace("\"", "\\\"");
    }

    protected String escapeDoubleQuotes(final Object obj) {
        return escapeDoubleQuotes(String.valueOf(obj));
    }
```

Reply:
They are not new methods, they existed in the DOTCFGVisualizer. We had all of these three methods before. The bodies of `prepareString` and `escapeDoubleQuotes` are the same.
`escapeDoubleQuotes` is used in the methods about visualizeStore().

Cool, it sounds great! I will overload the name.